pygbm/grower.py

```diff
@@ -82,18 +95,65 @@ def grow(self):
     def _intilialize_root(self):
         n_samples = self.features_data.shape[0]
         depth = 0
+        if self.splitter.constant_hessian:
+            hessian = self.splitter.all_hessians[0] * n_samples
+        else:
+            hessian = self.splitter.all_hessians.sum()
         self.root = TreeNode(depth, np.arange(n_samples, dtype=np.uint32),
                              self.splitter.all_gradients.sum(),
-                             self.splitter.all_hessians.sum())
+                             hessian)
         if (self.max_leaf_nodes is not None and self.max_leaf_nodes == 1):
             self._finalize_leaf(self.root)
             return
         self._compute_spittability(self.root)
 
-    def _compute_spittability(self, node):
-        split_info = self.splitter.find_node_split(node.sample_indices)
-        node.split_info = split_info
-        if split_info.gain < self.min_gain_to_split:
+    def _compute_spittability(self, node, only_hist=False):
+        """Compute histograms and split_info of a node and either make it a
+        leave or push it on the splittable node heap.
+
+        only_hist is used when _compute_spittability was called by a sibling
+        node: we only want to compute the histograms, not finalize or push
+        the node. If _compute_spittability is called again by the grower on
+        this same node, the histograms won't be computed again.
+        """
+
+        # Compute split_info and histograms if not already done
+        if node.split_info is None and node.histograms is None:
+            # If the sibling has less samples, compute its hist first (with
+            # the regular method) and use the subtraction method for the
+            # current node
+            if node.sibling is not None:  # root has no sibling
+                n_samples_sibling = node.sibling.sample_indices.shape[0]
+                n_samples_node = node.sample_indices.shape[0]
+                if n_samples_sibling < n_samples_node:
+                    self._compute_spittability(node.sibling, only_hist=True)
+                    # As hist of sibling is now computed we'll use the hist
+                    # subtraction method for the current node.
+                    node.hist_subtraction = True
+
+            tic = time()
+            if node.hist_subtraction:
+                split_info, histograms = (
+                    self.splitter.find_node_split_subtraction(
+                        node.sample_indices, node.parent.histograms,
+                        node.sibling.histograms)
+                )
+            else:
+                split_info, histograms = self.splitter.find_node_split(
+                    node.sample_indices)
+            toc = time()
+            node.time = toc - tic
```

Comment:
I think we should also accumulate the the total time spent finding the best splits in a grower level attribute and the same for the time spent in applying the splits.

In the gradient boosting class we should also record the time spent in binning and finally in computing the predictions to get the new gradients and hessians for the next boosting iteration.

At the end of the fit of the gradient boosting, if verbose > 0 we should print a profiling report with the accumulated time for each of those type of operations in a similar way as LightGBM does.

Reply:
Got it

There's a lot of

```python
tic = time()
blablah()
toc = time()
time_spent = toc - tic
```

Do we want a context manager that would allow us to do something like

```python
with Timer() as time_spent:
    blahblah()
```

?

Reply:
As you wish but I think it's fine like it is now. Using a context manager will add an indentation level.

Reply:
Ok let's keep it this way then